### PR TITLE
feat(rc): warn before publishing a post the account cannot afford

### LINKED ISSUE #3496: Warn before publishing a post the account cannot afford in RC
Follow-up to #3494, which fixed the offer *after* a failure. The app still has no RC estimate at all, so nothing warns beforehand.

## Why it matters

RC cost is dominated by serialized transaction size, so a long post can cost many times an average comment. Someone can write for twenty minutes, hit publish, and only then learn the account cannot pay. In the case that started this work the post needed 23.3B RC against an account whose *maximum* was 21.4B, so waiting for regeneration would never have worked either. Nothing in the app could have said so in advance.

## What to add

A non-blocking warning in the editor, shown while the draft is being written, when the estimated cost exceeds available RC. Tapping it opens the existing two-route offer sheet from #3494, so there is one place that offers a top-up or a boost.

`@ecency/sdk` 2.3.85 ships `estimateRcPrecheck`, which prices the actual operation the way the chain does rather than using the network average. Mobile is on `^2.3.83` and needs the bump.

Two details that matter for correctness, learned on the web side (ecency/vision-web#1489):

- The estimate must be built from the payload that is actually broadcast. Pricing the raw editor draft with tags-only metadata understates a post carrying a summary, images, links or a poll, and understating is the one direction that lets the chain reject a post we called affordable.
- Without a payload the estimator prices a *minimal* operation, which is cheaper than the network average and so flatters the number. That fallback is a lower bound, correct for a warning because it can never invent one, but it is not a number to display as "posts you can afford".

## Not in scope

The RC bar and its affordability counts. Mobile does not have that surface today, so there is nothing to reconcile.

### PR DESCRIPTION
**Pull request by feruzm**
Closes #3496. Follow-up to #3495, which fixed the offer *after* a failure.

## Why

The first anyone heard of an RC shortfall was the chain rejecting a post they had already finished writing. Worse, for a large enough post waiting does not help: cost is dominated by serialized transaction size, and in the report that started this work it reached 23.3B RC against an account whose *maximum* was 21.4B. No amount of regeneration would have made that post publishable, and nothing in the app could say so.

## What

A non-blocking warning in the editor, shown while the draft is being written, when the estimated cost exceeds available RC. Tapping it opens the two-route offer sheet from #3495, so there is still one place that sells a top-up or a boost.

**`buildEditorRcPayload`** (`src/utils/rcPayload.ts`) assembles the operation the editor actually broadcasts: the same permlink shape, parent, tag fallback and full `makeJsonMetadata` output. Pricing the raw draft with tags-only metadata would understate a post carrying a summary, images or links, and understating is the one direction that lets the chain reject a post we called affordable.

Two deliberate departures, both in the safe direction:

- `fetchRatios` is off. Fetching image ratios would put requests on the network every time typing pauses. The few bytes they add can only make the estimate lower, never a false alarm.
- The reply permlink is generated by the caller rather than per rebuild, since it is time-derived and only its length feeds the estimate.

Costing parses the body, so it runs 800ms after typing stops rather than on every keystroke, and a draft that cannot be costed shows nothing rather than guessing.

**`useRcPrecheck`** (`src/hooks/useRcPrecheck.ts`) wraps `estimateRcPrecheck` over the three SDK queries. Without a payload the estimator prices a *minimal* operation, a lower bound that can miss a marginal case but never invents a warning, which is the right default for this.

## SDK bump

`^2.3.83` to `^2.3.85` for `estimateRcPrecheck`, which prices the actual operation the way the chain does rather than using the network average. The average is dominated by short replies: it told the account in that report it could afford 17 more posts when the next one needed more RC than the account could ever hold. Lockfile diff is the single `@ecency/sdk` entry.

## Testing

`src/utils/rcPayload.test.ts`, 10 cases: post and reply shapes, the five-word permlink cap, the `hive-125125` tag fallback, blank-tag filtering, AI disclosure, metadata size against a tags-only baseline, and the two nothing-to-price cases.

Full suite 855 (10 new), typecheck clean at 0 errors, lint identical to `development`.

## Device check

The estimator's arithmetic is covered by the SDK's own tests and was validated against a real rejection there, but the banner appearing at the right moment is worth confirming on a low-RC account: type past the threshold, confirm it appears, confirm tapping it opens the sheet, and confirm it does not appear for a normal short post.

## Summary by CodeRabbit

* **New Features**
  * Added a pre-publish warning that estimates Resource Credit requirements for posts, replies, and edits.
  * Supports estimates based on draft size, metadata, tags, media, polls, rewards, and beneficiaries.
  * Users can open the Resource Credit offer sheet when a draft may exceed available credits.
* **Localization**
  * Added English messaging for drafts that may be too large to publish.
* **Tests**
  * Added coverage for Resource Credit payload generation and editor edge cases.

### CODE REVIEW

File: src/utils/rcPayload.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  const jsonMetadata = makeJsonMetadata(
+    meta,
+    isReply ? post?.json_metadata?.tags || ['ecency'] : tags,
+  );
+
+  return {
+    kind: 'comment',
+    op: {
+      author: username,
+      permlink: isReply ? replyPermlink ?? '' : generatePermlink(title),
```
[the remark this answers is not in the document]
Comment (feruzm) [reply src/utils/rcPayload.ts]: Fixed in 618f2878da. Confirmed and worse than a near-threshold problem: `makeOptions` returns a full options object whenever author and permlink are present, so a post sends comment_options on default reward settings too. Every post was priced short by a whole operation plus its beneficiaries, not just the customised ones.

The payload now carries `options` for posts and omits it for replies, which genuinely send none. Three tests: options present on defaults, the real beneficiaries carried through, and absent for a reply.

File: src/screens/editor/screen/editorScreen.tsx [hunk 2]
```diff
@@ -569,6 +570,12 @@ class EditorScreen extends Component<any, any> {
           isPreviewActive={isPreviewActive}
         > */}
         <Fragment>
+          <RcPrecheckBanner
+            username={currentAccount?.name}
+            fields={fields}
+            post={post}
+            isReply={isReply}
```
[the remark this answers is not in the document]
Comment (feruzm) [reply src/screens/editor/screen/editorScreen.tsx:580]: Fixed in 618f2878da, by modelling edits rather than hiding the banner.

Checking the update path first was worth it, because the divergence is bigger than the permlink: `editorContainer` sends `createPatch(oldBody, newBody)` whenever the patch is smaller than the old body, keeps the original permlink and parent, and merges metadata through `makeJsonMetadataForUpdate`. Pricing that as a new post overstates a one-word fix to a long article by the entire article, so it invents warnings.

I chose to build the real edit payload because an edit can genuinely exhaust RC and suppressing the banner would leave that case silent. Six tests cover it: original identity kept, a small change to a long post staying small, the whole body sent when a diff would not be smaller, no comment_options, an untouched AI disclosure surviving the merge, and nothing returned when there is no post.

File: src/utils/rcPayload.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  if (!username || !body) {
+    return undefined;
+  }
+
+  const title = fields?.title ?? '';
+  const meta = await extractMetadata({
+    body,
+    fetchRatios: false,
+    ...(isReply ? { postType: PostTypes.COMMENT } : {}),
+  });
```
[the remark this answers is not in the document]
Comment (feruzm) [reply src/utils/rcPayload.ts:117]: Fixed in 618f2878da. All three inputs now come through: `pollDraft`, `videoThumbUrls` and `thumbUrl` are passed to `extractMetadata` exactly as the submit path passes them, and edit mode builds the real update payload rather than deriving fresh identifiers.

A poll is the clearest case: it lands in json_metadata via `convertToPollMeta` and a twelve-choice poll is not a rounding error, so a poll post could pass the check and then be rejected, which is the failure this feature exists to prevent. Tests assert the metadata grows with a poll and with video thumbnails, and the edit behaviour is covered separately.

File: src/utils/rcPayload.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  const aiTools = cleanAiTools(fields?.aiTools);
+  if (aiTools) {
+    meta.ai_tools = aiTools;
+  }
+
+  const tags = (fields?.tags ?? []).filter((tag: any) => tag && tag !== ' ');
+  const jsonMetadata = makeJsonMetadata(
+    meta,
+    isReply ? post?.json_metadata?.tags || ['ecency'] : tags,
+  );
```
[the remark this answers is not in the document]
Comment (feruzm) [reply src/utils/rcPayload.ts:128]: Fixed in 618f2878da for the inputs that reach the broadcast: `thumbUrl`, `videoThumbUrls` and `pollDraft` are now threaded through, and comment_options carries the reward type and beneficiaries.

One correction on the specifics. `fields.meta` at editorScreen 353-369 is the draft-persistence snapshot; the broadcast metadata is rebuilt from scratch in `_submitPost` by calling `extractMetadata` with those same raw inputs. So reusing `fields.meta` would not be reusing the broadcast serializer, it would be reading a parallel copy. Feeding the same inputs to the same function is what actually keeps the two in step, which is what this now does.

Description and reward type are worth separating too: reward type shapes comment_options, which is now priced, while `postDescription` reaches metadata through `extractMetadata`'s own summary handling rather than as a separate field.

File: src/components/rcPrecheckBanner/rcPrecheckBanner.tsx [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+          }
+        });
+    }, BUILD_DEBOUNCE_MS);
+
+    return () => {
+      cancelled = true;
+      clearTimeout(timer);
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [username, title, body, JSON.stringify(tags), post?.author, post?.permlink, isReply]);
```
[the remark this answers is not in the document]
Comment (feruzm) [reply src/components/rcPrecheckBanner/rcPrecheckBanner.tsx]: Fixed in 618f2878da. Correct: `_handleAiToolUsed` changes only that field, so a disclosure toggled after the last title, body or tag edit left an obsolete estimate standing.

`JSON.stringify(fields?.aiTools)` is in the dependency list now, along with beneficiaries, poll draft, video thumbnails, thumbnail, reward type and the edited post's body. They are serialized rather than listed by reference because the editor hands these down as fresh objects on every render, which would restart the debounce forever.

[the remark this answers is not in the document]
Comment (feruzm) [reply src/components/rcPrecheckBanner/rcPrecheckBanner.tsx]: Fixed in 618f2878da for the values that change the estimate: aiTools, beneficiaries, poll draft, video thumbnails, thumbnail, reward type and the edited post's markdown body are all in the dependency list now.

Keeping the disable is deliberate rather than an oversight. `fields` and `post` are recreated on every editor render, so listing them by reference would restart the 800ms debounce on each keystroke and the estimate would never settle. The list holds serialized values of exactly the fields that feed the payload, which is the same contract the rule is asking for, expressed in a way that terminates.

File: src/components/rcPrecheckBanner/rcPrecheckBanner.tsx [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    const timer = setTimeout(() => {
+      buildEditorRcPayload({
+        username: username!,
+        fields,
+        post,
+        isReply,
+        // Time-derived at broadcast, so only its length feeds the estimate.
+        replyPermlink: isReply
+          ? generateUniquePermlink(`re-${String(post?.author ?? '').replace(/\./g, '')}`)
+          : undefined,
```
[the remark this answers is not in the document]
Comment (feruzm) [reply src/components/rcPrecheckBanner/rcPrecheckBanner.tsx]: Fixed in 618f2878da. `replyPermlink` is now memoized on the parent author rather than regenerated inside the debounced rebuild.

You are right that it matters beyond tidiness: the permlink is time-derived, so only its length reaches the estimate, and the millisecond component changes that length at digit boundaries. That made the priced size jitter by a byte or two for no reason.

File: src/utils/rcPayload.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (feruzm) [reply src/utils/rcPayload.ts:117]: Fixed in b44c4a89ac.

Right, and the stored list understated exactly the posts most likely to sit near the line. Two additions happen at submit time.

The 3Speak row is exact now: the builder calls the same `enforceThreeSpeakBeneficiary` the submit path calls, on the same body, so an embedded video adds the same mandatory row to the priced comment_options.

The support row needed a judgement call. Submit resolves it with a decrypted access token and a network fetch, which is far too heavy to repeat while someone types. It now comes from the shared query cache instead: `getSupportSettingsQueryOptions` is keyed `["support","settings",username]` with no token in the key, so the settings screen and the beneficiary modal warm the same entry and a cold cache costs nothing. When the cache is cold the row is skipped, which leaves the estimate lower by one beneficiary rather than raising a warning that would not have happened. The container now exposes whether the author set an explicit list, since that is the condition submit uses to decide on the row at all.

Five tests: the 3Speak row added by an embed, the support row added when no list was set, an author-set list left alone, nothing added for the support account itself, and nothing added when the percentage is unknown.
